<a href="https://colab.research.google.com/github/Guhan2348519/dl-lab-tasks/blob/main/2348519_dl_lab9.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [ ]:
/content/drive/MyDrive/Tweets.csv

In [ ]:
import pandas as pd
import numpy as np
import tensorflow as tf
from tensorflow.keras.preprocessing.text import Tokenizer
from tensorflow.keras.preprocessing.sequence import pad_sequences
from sklearn.model_selection import train_test_split

In [ ]:
df=pd.read_csv('/content/drive/MyDrive/Tweets.csv')
df.head(5)

tweet_id airline_sentiment  airline_sentiment_confidence  \
0  5.703060e+17           neutral                        1.0000   
1  5.703010e+17          positive                        0.3486   
2  5.703010e+17           neutral                        0.6837   
3  5.703010e+17          negative                        1.0000   
4  5.703010e+17          negative                        1.0000   

  negativereason  negativereason_confidence         airline  \
0            NaN                        NaN  Virgin America   
1            NaN                     0.0000  Virgin America   
2            NaN                        NaN  Virgin America   
3     Bad Flight                     0.7033  Virgin America   
4     Can't Tell                     1.0000  Virgin America   

  airline_sentiment_gold        name negativereason_gold  retweet_count  \
0                    NaN     cairdin                 NaN              0   
1                    NaN    jnardino                 NaN              0   
2                    NaN  yvonnalynn                 NaN              0   
3                    NaN    jnardino                 NaN              0   
4                    NaN    jnardino                 NaN              0   

                                                text tweet_coord  \
0                @VirginAmerica What @dhepburn said.         NaN   
1  @VirginAmerica plus you've added commercials t...         NaN   
2  @VirginAmerica I didn't today... Must mean I n...         NaN   
3  @VirginAmerica it's really aggressive to blast...         NaN   
4  @VirginAmerica and it's a really big bad thing...         NaN   

      tweet_created tweet_location               user_timezone  
0  24-02-2015 11:35            NaN  Eastern Time (US & Canada)  
1  24-02-2015 11:15            NaN  Pacific Time (US & Canada)  
2  24-02-2015 11:15      Lets Play  Central Time (US & Canada)  
3  24-02-2015 11:15            NaN  Pacific Time (US & Canada)  
4  24-02-2015 11:14            NaN  Pacific Time (US & Canada)

In [ ]:
actualSentiment = df['airline_sentiment']
positiveNegative = actualSentiment.map(lambda x: 1 if x=='positive' else 0)
df['Sentiment'] = positiveNegative
df['Sentiment'].value_counts()

Sentiment
0    12277
1     2363
Name: count, dtype: int64

In [ ]:
# Preprocessing
# Assuming 'text' column contains the tweets and 'airline_sentiment' column contains sentiment labels
texts = df['text'].astype(str)
labels = df['airline_sentiment']
maxlen = 100  # Max length of each sequence
max_words = 10000  # Vocabulary size

tokenizer = Tokenizer(num_words=max_words)
tokenizer.fit_on_texts(texts)
sequences = tokenizer.texts_to_sequences(texts)

word_index = tokenizer.word_index
print('Found %s unique tokens.' % len(word_index))

data = pad_sequences(sequences, maxlen=maxlen)

labels = pd.get_dummies(labels)

X_train, X_test, y_train, y_test = train_test_split(data, labels, test_size=0.2, random_state=42)

Found 15768 unique tokens.


In [ ]:
# RNN Model
model_rnn = tf.keras.models.Sequential([
    tf.keras.layers.Embedding(max_words, 32),
    tf.keras.layers.SimpleRNN(32),
    tf.keras.layers.Dense(3, activation='softmax')  # 3 output classes: negative, neutral, positive
])

model_rnn.compile(optimizer='adam',
              loss='categorical_crossentropy',
              metrics=['accuracy'])

print(model_rnn.summary())

history_rnn = model_rnn.fit(X_train, y_train,
                    epochs=10,
                    batch_size=32,
                    validation_split=0.2)  # 20% of training data for validation

Model: "sequential"
_________________________________________________________________
 Layer (type)                Output Shape              Param #   
 embedding (Embedding)       (None, None, 32)          320000    
                                                                 
 simple_rnn (SimpleRNN)      (None, 32)                2080      
                                                                 
 dense (Dense)               (None, 3)                 99        
                                                                 
Total params: 322179 (1.23 MB)
Trainable params: 322179 (1.23 MB)
Non-trainable params: 0 (0.00 Byte)
_________________________________________________________________
None
Epoch 1/10
293/293 [==============================] - 17s 49ms/step - loss: 0.8054 - accuracy: 0.6635 - val_loss: 0.6522 - val_accuracy: 0.7328
Epoch 2/10
293/293 [==============================] - 8s 26ms/step - loss: 0.4886 - accuracy: 0.8144 - val_loss: 0.6173 - val_accuracy:

This code creates an RNN model with an embedding layer, a SimpleRNN layer, and a dense output layer with a softmax activation function. It compiles the model using the Adam optimizer and categorical cross-entropy loss function, then trains it for 10 epochs on the training data with a validation split of 20%.

In [ ]:
# LSTM Model
model_lstm = tf.keras.models.Sequential([
    tf.keras.layers.Embedding(max_words, 32),
    tf.keras.layers.LSTM(32),
    tf.keras.layers.Dense(3, activation='softmax')
])

model_lstm.compile(optimizer='adam',
              loss='categorical_crossentropy',
              metrics=['accuracy'])

print(model_lstm.summary())

history_lstm = model_lstm.fit(X_train, y_train,
                    epochs=10,
                    batch_size=32,
                    validation_split=0.2)

Model: "sequential_1"
_________________________________________________________________
 Layer (type)                Output Shape              Param #   
 embedding_1 (Embedding)     (None, None, 32)          320000    
                                                                 
 lstm (LSTM)                 (None, 32)                8320      
                                                                 
 dense_1 (Dense)             (None, 3)                 99        
                                                                 
Total params: 328419 (1.25 MB)
Trainable params: 328419 (1.25 MB)
Non-trainable params: 0 (0.00 Byte)
_________________________________________________________________
None
Epoch 1/10
293/293 [==============================] - 17s 50ms/step - loss: 0.7348 - accuracy: 0.6890 - val_loss: 0.6280 - val_accuracy: 0.7384
Epoch 2/10
293/293 [==============================] - 14s 48ms/step - loss: 0.4809 - accuracy: 0.8078 - val_loss: 0.5607 - val_accura

You created an LSTM model with embedding and dense layers, compiled it with Adam optimizer, and trained it for 10 epochs on training data with a validation split of 20%.LSTM (Long Short-Term Memory) is a type of recurrent neural network (RNN) architecture commonly used for processing and making predictions based on sequential data. It is designed to capture long-term dependencies and overcome the limitations of traditional RNNs in handling information over extended time lags.


In [ ]:
# Evaluate the models
loss_rnn, accuracy_rnn = model_rnn.evaluate(X_test, y_test)
print("RNN Test Accuracy:", accuracy_rnn)

loss_lstm, accuracy_lstm = model_lstm.evaluate(X_test, y_test)
print("LSTM Test Accuracy:", accuracy_lstm)

92/92 [==============================] - 1s 7ms/step - loss: 1.3320 - accuracy: 0.7131
RNN Test Accuracy: 0.7131147384643555
92/92 [==============================] - 1s 11ms/step - loss: 1.0363 - accuracy: 0.7691
LSTM Test Accuracy: 0.7691256999969482


The RNN model attained a test accuracy of approximately 71.31%, while the LSTM model achieved a slightly higher test accuracy of about 76.91%. Despite their differences in accuracy, both models demonstrated relatively similar performance on the test dataset.

In [ ]:

import random

# Function to predict sentiment using the specified model and tokenizer
def predict_sentiment(text, model, tokenizer):
    tw = tokenizer.texts_to_sequences([text])  # Remove the outer list
    tw = pad_sequences(tw, maxlen=100)  # Assuming max length is 100
    prediction = model.predict(tw)  # No need to round or convert to int yet
    return prediction

# Randomly select a text from the dataset
random_index = random.randint(0, len(df) - 1)
random_text = df['text'][random_index]

# Make predictions using the RNN and LSTM models
rnn_prediction = predict_sentiment(random_text, model_rnn, tokenizer)
lstm_prediction = predict_sentiment(random_text, model_lstm, tokenizer)

# Convert predictions to labels
rnn_label = "Positive" if rnn_prediction.argmax() == 1 else "Negative"
lstm_label = "Positive" if lstm_prediction.argmax() == 1 else "Negative"

# Print the predictions
print("Random Text:")
print(random_text)
print("\nRNN Model Predicted Sentiment:", rnn_label)
print("LSTM Model Predicted Sentiment:", lstm_label)

1/1 [==============================] - 1s 510ms/step
Random Text:
@SouthwestAir: grrrrrr I need a flight there first! Maybe for my bday next month...

RNN Model Predicted Sentiment: Positive
LSTM Model Predicted Sentiment: Positive
